esteid/actions.py

```diff
@@ -1,86 +1,178 @@
+
+        if not files:
             return {
                 'success': False,
-                'error_code': e.error_code,
-                'message': service.ERROR_CODES.get(int(e.error_code), service.ERROR_CODES[100])
+                'code': 'MIN_1_FILE',
             }
 
+        container = open_container(container_path, files)
+        xml_sig = prepare_signature(certificate, container)
 
-class IdCardFinishAction(BaseAction):
-    @classmethod
-    def do_action(cls, view, action_kwargs):
-        service = view.get_service()
+        # save intermediate signature XML to temp file
+        with NamedTemporaryFile(delete=False) as f:
+            f.write(xml_sig.dump())
 
-        try:
-            # Should we also pass back the signed_doc_info (?)
-            service.finalize_signature(**action_kwargs)
+        if not container.name:
+            with NamedTemporaryFile(delete=False) as temp_container_file:
+                container.save(temp_container_file.name)
+        else:
+            temp_container_file = None
+
+        signed_digest = xml_sig.digest()
+        digest_hash_b64 = binascii.b2a_base64(signed_digest).decode()
+
+        update_esteid_session(
+            request,
+            signed_hash=digest_hash_b64,  # probably not needed, as we can take the hash from the signature XML
+            temp_signature_file=f.name,
+            temp_container_file=temp_container_file.name if temp_container_file else None,
+        )
+
+        logger.debug("Signature hash base64: %s", digest_hash_b64)
+
+        return {
+            'success': True,
+            'digest': binascii.b2a_hex(signed_digest).decode(),
+        }
 
-            return {
-                'success': True,
-            }
 
-        except DigiDocNotOk as e:
+class IdCardFinishAction(BaseAction):
+    @classmethod
+    def do_action(cls, view: "GenericDigitalSignViewMixin", params: dict = None, *, signature_value: str = None,
+                  container_path: str = None):
+        """
+        The old API is to pass a dict of params (previously confusingly named `action_kwargs`),
+        the keyword args are added here for clarity as to what the method accepts
+
+        :param view:
+        :param params:
+        :param signature_value: a HEX encoded signature, as received from `hwcrypto.js`
+        :return:
+        """
+        request = view.request
+        session_data = get_esteid_session(request)
+        if not session_data:
             return {
                 'success': False,
-                'error_code': e.code,
-                'message': str(e),
+                'code': 'NO_SESSION',
             }
 
-        except DigiDocError as e:
+        if signature_value is None:
+            signature_value = params['signature_value']
+            if not signature_value:
+                return {
+                    'success': False,
+                    'code': 'BAD_SIGNATURE',
+                }
+
+        if container_path is None:
```

Comment:
Ditto about where this comes from (context: security and prevention of opening unwanted files).